In [1]:
# import dependencies
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [2]:
# import sklearn metrics
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import r2_score
from imblearn.metrics import classification_report_imbalanced

In [3]:
# load the data
merged_df = pd.read_csv("merged_df.csv", index_col=[0],sep=',')
merged_df.head()

GeoFips      county  pcp_count  population  pop_density  per_capita_income  \
0    12086  Miami-Dade      13619     2715516   552.187071              57213   
1     6025    Imperial        236      180216    16.659991              44500   
2    40109    Oklahoma       2914      782051   425.989528              56971   
3    39115      Morgan         19       14702    13.631412              38583   
4     8059   Jefferson       1459      570427   288.167016              68829   

   pcp_per_capita  pop_density_lvl  region_far_west  region_great_lakes  ...  \
0        5.015253                3                0                   0  ...   
1        1.309540                1                1                   0  ...   
2        3.726100                3                0                   0  ...   
3        1.292341                1                0                   1  ...   
4        2.557733                3                0                   0  ...   

   state_TN  state_TX  state_UT  state_VA  state_VT  state_WA  state_WI  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   state_WV  state_WY  pcp_per_capita_bins  
0         0         0                    1  
1         0         0                    0  
2         0         0                    1  
3         0         0                    0  
4         0         0                    1  

[5 rows x 68 columns]

In [4]:
# Create features
X = merged_df.drop(columns=["pcp_per_capita", "pcp_per_capita_bins", "pop_density"])
X = pd.get_dummies(X) 

# Create target
y = merged_df["pcp_per_capita_bins"]

In [5]:
X.describe()

GeoFips     pcp_count    population  per_capita_income  \
count   3029.000000   3029.000000  3.029000e+03        3029.000000   
mean   29981.758666    339.705513  1.141870e+05       48949.537801   
std    15070.955336   1094.505146  4.231330e+05       13078.053879   
min     1001.000000      1.000000  4.590000e+02       22644.000000   
25%    18153.000000     18.000000  1.196200e+04       40971.000000   
50%    29095.000000     56.000000  2.721600e+04       46571.000000   
75%    44001.000000    198.000000  7.152200e+04       53791.000000   
max    56045.000000  18526.000000  1.009805e+07      220645.000000   

       pop_density_lvl  region_far_west  region_great_lakes  region_mideast  \
count      3029.000000      3029.000000         3029.000000     3029.000000   
mean          1.496864         0.059756            0.144272        0.059426   
std           1.118362         0.237073            0.351423        0.236458   
min           0.000000         0.000000            0.000000        0.000000   
25%           0.000000         0.000000            0.000000        0.000000   
50%           1.000000         0.000000            0.000000        0.000000   
75%           2.000000         0.000000            0.000000        0.000000   
max           3.000000         1.000000            1.000000        1.000000   

       region_new_england  region_plains  ...  county_Yellowstone  \
count         3029.000000    3029.000000  ...          3029.00000   
mean             0.022450       0.193793  ...             0.00033   
std              0.148165       0.395334  ...             0.01817   
min              0.000000       0.000000  ...             0.00000   
25%              0.000000       0.000000  ...             0.00000   
50%              0.000000       0.000000  ...             0.00000   
75%              0.000000       0.000000  ...             0.00000   
max              1.000000       1.000000  ...             1.00000   

       county_Yoakum  county_Yolo  county_York  county_Young  county_Yuba  \
count     3029.00000   3029.00000  3029.000000    3029.00000   3029.00000   
mean         0.00033      0.00033     0.001321       0.00033      0.00033   
std          0.01817      0.01817     0.036322       0.01817      0.01817   
min          0.00000      0.00000     0.000000       0.00000      0.00000   
25%          0.00000      0.00000     0.000000       0.00000      0.00000   
50%          0.00000      0.00000     0.000000       0.00000      0.00000   
75%          0.00000      0.00000     0.000000       0.00000      0.00000   
max          1.00000      1.00000     1.000000       1.00000      1.00000   

       county_Yukon-Koyukuk  county_Yuma  county_Zapata  county_Zavala  
count            3029.00000  3029.000000     3029.00000     3029.00000  
mean                0.00033     0.000660        0.00033        0.00033  
std                 0.01817     0.025692        0.01817        0.01817  
min                 0.00000     0.000000        0.00000        0.00000  
25%                 0.00000     0.000000        0.00000        0.00000  
50%                 0.00000     0.000000        0.00000        0.00000  
75%                 0.00000     0.000000        0.00000        0.00000  
max                 1.00000     1.000000        1.00000        1.00000  

[8 rows x 1812 columns]

In [6]:
# Check the balance of our target values
y.value_counts()

0    1519
1    1510
Name: pcp_per_capita_bins, dtype: int64

In [7]:
# Train, test, split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,
    y, random_state=1)

Change estimators from 100 to 200

In [8]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
random_forest = BalancedRandomForestClassifier(n_estimators = 200)
random_forest = random_forest.fit(X_train, y_train)

In [9]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = random_forest.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.8271972819427426

In [10]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(y_test, y_pred)
print(matrix)

[[312  64]
 [ 67 315]]


In [11]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.82      0.83      0.82      0.83      0.83      0.68       376
          1       0.83      0.82      0.83      0.83      0.83      0.68       382

avg / total       0.83      0.83      0.83      0.83      0.83      0.68       758



In [12]:
# List the features sorted in descending order by feature importance
feature_names = X.columns
sorted(zip(random_forest.feature_importances_, feature_names), reverse=True)

[(0.229490103828629, 'pcp_count'),
 (0.11853523533537698, 'population'),
 (0.08808991120056384, 'per_capita_income'),
 (0.07141984812432739, 'GeoFips'),
 (0.033233621826644515, 'pop_density_lvl'),
 (0.007952558124544191, 'region_southeast'),
 (0.00788942174746801, 'state_TX'),
 (0.007305464770303128, 'region_southwest'),
 (0.004653560786430523, 'state_CA'),
 (0.004610770715139313, 'region_plains'),
 (0.004224519778834199, 'region_mideast'),
 (0.0038815989091070184, 'region_great_lakes'),
 (0.0038185889919781016, 'state_KY'),
 (0.003808177210068617, 'state_GA'),
 (0.0036597775941273943, 'region_new_england'),
 (0.003482082816032121, 'region_far_west'),
 (0.003074100279706001, 'state_AL'),
 (0.0027906757716978227, 'state_KS'),
 (0.0027664779884672215, 'state_MO'),
 (0.0027329256587837287, 'state_OK'),
 (0.0026887310149447236, 'county_Union'),
 (0.002683965767486249, 'region_rocky_mountain'),
 (0.00259884238895343, 'state_MN'),
 (0.0023706557252782343, 'county_Jefferson'),
 (0.00234437336

Change estimators from 100 to 200

In [13]:
# Train the EasyEnsembleClassifier
from imblearn.ensemble import EasyEnsembleClassifier
easy = EasyEnsembleClassifier(n_estimators = 200,random_state=1)
easy = easy.fit(X_train, y_train)

In [14]:
# Calculated the balanced accuracy score
y_pred = easy.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.8970285173220452

In [15]:
# Display the confusion matrix
y_pred = easy.predict(X_test)
matrix = confusion_matrix(y_test, y_pred)
print(matrix)

[[334  42]
 [ 36 346]]


In [16]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.90      0.89      0.91      0.90      0.90      0.80       376
          1       0.89      0.91      0.89      0.90      0.90      0.81       382

avg / total       0.90      0.90      0.90      0.90      0.90      0.80       758



In [17]:
# find r-squared score
r2_score(y_test, y_pred)

0.588364709813969

After hyper tuning using twice as many estimators, the accuracy went from 89.6% to 89.7% and the r2 score went from 58.3% to 58.9%.